### DISCUSSION TITLE
News outside of mapme.biodiversity that can be of interest

### DISCUSSION THREAD

**Discussion by fBedecarrats** [mapme-initiative/mapme.biodiversity#187] [General]
I propose to open this thread so all contributors can post links to news that can be inspiring for the development of this package and the underlying framework.

**Comment by fBedecarrats** [comment]
@cboettig recently posted about the [Source Cooperative](https://source.coop/) in #186

**Comment by fBedecarrats** [comment]
This video showcases the use of wind data from the ERA5 dataset using [the ecmwfr package](https://github.com/bluegreen-labs/ecmwfr)

**Comment by cboettig** [comment]
Thanks!   In this spirit, have you seen the [duckdb spatial extension](https://github.com/duckdblabs/duckdb_spatial) ?  A lot of remote assets are in tabular form with lat/lon columns in csv etc.  I've been exploring how we can use duckdb to access and perform spatial operations on those data without downloading them.  

Here's a trivial example:

```r
library(duckdbfs)
library(dplyr)
spatial_csv <- paste0("https://raw.githubusercontent.com/cboettig/duckdbfs/",
                     "7189ed7e0456927a55a877acf77aaadc074ba01b/",
                     "inst/extdata/spatial-test.csv") |>
  open_dataset(format = "csv") 

spatial_csv |>
  mutate(geometry = ST_Point(longitude, latitude)) |>
  to_sf()
  ```


For an ambitious example, here I try and do a spatial polygon filter on the GBIF data on AWS.  This takes about 18min on my machine, but even a non-spatial filter, e.g. by bounding box, takes about 8min since GBIF is quite large.  (This uses my thin wrapper around `duckdb` to streamline some of the syntax).


```r
library(dplyr)
library(sf)
library(spData)
library(duckdbfs) #  remotes::install_github("cboettig/duckdbfs") 

# let's say we have some polygon to filter by:
costa_rica <- world |> filter(grepl("Costa Rica", name_long)) |> pull(geom) |> st_as_text()

# connect to the dataset
gbif <- duckdbfs::open_dataset("s3://gbif-open-data-us-east-1/occurrence/2022-12-01/occurrence.parquet", s3_region="us-east-1")

# Create a geometry column from the point data, then use it in filters
cr_species <- 
  gbif |> 
  mutate(geometry = ST_Point(decimallongitude, decimallatitude)) |>  
  filter(class == "Mammalia") |>
  filter(ST_Within(geometry, ST_GeomFromText({costa_rica}))) |>
  to_sf()
```

Haven't tested this that extensively, but curious what you think and if it could be relevant here.  The docs for the spatial extension have richer examples and notes on all the [supported spatial functions](https://github.com/duckdblabs/duckdb_spatial#supported-functions)

**Comment by goergen95** [comment]
[Iconnect](https://CRAN.R-project.org/package=lconnect) is an R package to analyse landscape connectivity that also works with vector data.

**Comment by goergen95** [comment]
[Chelsea](https://chelsa-climate.org/) is a high-resolution (~1km) climatology data set with a very high number of pre-computed variables.

**Reply by goergen95** [reply]
Added in d2986fe3ebfcb6b759842455d698ebdbdeaba619

**Comment by goergen95** [comment]
[WorldCereal](https://essd.copernicus.org/articles/15/5491/2023/), a dynamic, 10m crop and irrigation  product developed by ESA. Might be interesting for agricultural use-cases.

**Reply by fBedecarrats** [reply]
That looks great! For now it seems that they only have 2021, right? I am a bit surprised by the agro-ecological zones that they used to stratify the analysis: they seem really coarse to me (eg. only one zone for Madagascar when there are at least 3 different climates with very disting croping calendars in Madagascar...).

**Comment by goergen95** [comment]
```
Clay is a foundational model of Earth. It uses a Vision Transformer architecture adapted to 
understand geospatial and temporal relations on Earth Observation data. The model is 
trained via Self-supervised learning (SSL) using a Masked Autoencoder (MAE) method.

The Clay model can be used in three main ways:

   - Generate semantic embeddings for any location and time.

   - Fine-tune the model for downstream tasks such as classification, regression, and generative tasks.

   - Use the model as a backbone for other models.
```

[https://clay-foundation.github.io/model/](https://clay-foundation.github.io/model/)

**Reply by goergen95** [reply]
[NAIP](https://source.coop/clay/clay-v1-5-naip-2) embeddings over continental US on source.coop

**Comment by goergen95** [comment]
ACLED released a conflict exposure tool based on WorldPop. Could be an inspiration for a new indicator with UCDP GED. https://www.worldpop.org/blog/conflict-exposure-a-new-tool-to-measure-the-impact-of-conflict/

**Reply by goergen95** [reply]
First draft added [here](https://github.com/mapme-initiative/mapme.biodiversity/pull/321). Feedback to this new indicators is very welcomed!

**Comment by goergen95** [comment]
[Copernicus Marine Data Store](https://data.marine.copernicus.eu/products)

**Comment by goergen95** [comment]
New portal with [WAPOR V3](https://data.apps.fao.org/wapor/) released.

**Comment by goergen95** [comment]
There is a new OGC Standard Working Group currently in the process of formation on the topic of [Geospatial Indicator Reporting](https://portal.ogc.org/files/107038) in the context of SDGs.

**Comment by goergen95** [comment]
Upcoming GDAL 3.9 will have a feature implemented that sorts geoparquet files by bounding-box for faster reads when spatially filtering (see [here](https://www.github.com/OSGeo/gdal/pull/9185)). Could be useful when evaluating performant formats/file organizations to translate vector resources to in the next months.

**Comment by goergen95** [comment]
New coastal mid-res digital terrain model with low errors [here](https://www.nature.com/articles/s41597-024-03091-9).

**Comment by goergen95** [comment]
JRC's [technical report](https://publications.jrc.ec.europa.eu/repository/handle/JRC136960) on global forest map for 2020.

**Comment by goergen95** [comment]
Near-real-time estimation of carbon dioxide and water flux from geostationary satellites from the [ALIVE](https://alive-abi.github.io/alive/index.html) project.

**Comment by goergen95** [comment]
1m resolution with global coverage of tree heights from Meta and WRI -> [here](https://registry.opendata.aws/dataforgood-fb-forests/)... Absolutely crazy!

**Reply by fBedecarrats** [reply]
I understand from you posts on Linkedin and other comments on bluesky that there are still questions about the reliability of this dataset. Do you have an opinion on this now?

**Reply by goergen95** [reply]
I have not validated the data myself, but there is an interesting analysis by Iain Woodhouse available [here](https://www.earthblox.io/blog/earth-blox-dataset-review-meta-global-canopy-height-1m).

There are two main take-aways from my POV:

- its a global dataset which comes with the usual caveats, i.e. the reported global accuracy does not uphold to ones expectations when comparing it with local data sets derived from e.g. airborne LiDAR or similar. However, it can provide valuable information for areas where no higher quality measurements are available or for continental/global comparisons.

- the canopy height layer was generated from image mosaics with very different time stamps between 2009 (!) - 2020. In the first release the temporal information was missing but in my understanding it is now available. However, the temporal distribution is quite patchy, especially in the tropics due to persistent cloud cover, so that properly harmonizing this data with e.g. GFW will be really challenging in my view.

**Comment by cboettig** [comment]
Hi friends, have you seen the beautiful work of Gilberto Camara and colleagues in the `sits` package and accompanying text? https://e-sensing.github.io/sitsbook/   .  I haven't delved too much into it yet but they appear to have wrapped around rstac and gdalcubes rather nicely.

**Reply by fBedecarrats** [reply]
Thanks @cboettig ! I understand that this as some kind of extension of the {stars} approach on data cubes, applied to landcover classification and with a strong focus on accessing raster data that sits on the cloud.

**Reply by edzer** [reply]
We're working on an `openEO` interface around the `sits` package, and will try to deploy it on the [CDSE](https://dataspace.copernicus.eu/).

**Comment by fBedecarrats** [comment]
This paper is not that new but I found it interesting regarding the appropriate scale of analysis to use for spatial impact evaluation: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0167945
Briefly, it says that smaller area of analysis is not always better and one should take into account the scale at which the process of interest is hapenning (eg. deforestation, land use change...) and the resolution of the covariates used for analysis.

**Reply by goergen95** [reply]
Hi,

this simulation study is gold, thanks for sharing!

I have a few thoughts in my head with some question marks of how to relate
this study to what we are trying to do here:

1) The simulation study assumes a grid-based analysis which corresponds 
   well with the honeycomb procedure of the conducted impact assessments. 
   I wonder if there are similar studies when the unit of analysis is vector
   based? 
  
2) Depending on the scale of analysis, significant bias can be introduced when
   treatment is represented as a discrete variable rather than on a continuous scale. 
   I wonder if this holds with our `exactextract` approach with which for most 
   variables we extract area-weighted values based on the overlap between a 
   vector with individual cells.
   
3) The main recommendation, in my understanding, goes something like 
   *think about the scale of the process you are modelling to decide on an 
   appropriate scale of analysis*. This does not translate well to 
   global-portfolio based analysis, does it? When we are looking at globally
   distributed PAs, I assume that there are very different variables and 
   relationships between those driving the underlying processes. Selecting
   a scale would assume that a) the process can be observed globally and b)
   the scale of this process does not change with space. What kind of processes 
   are we interested in where we can reasonably assume this?

**Comment by goergen95** [comment]
Access to zarr data cubes for weather data via [dynamical.org](http://dynamical.org)

**Reply by cboettig** [reply]
that is cool!  Wish we had better Zarr support in R, I at least still struggle with the sf mdim interface. the xarray syntax looks so nice!  Looks like at the moment these are mostly forecast products (GEFS).  Also of interest for weather data may be the https://open-meteo.com/  project.

**Reply by goergen95** [reply]
Yes, xarrays zarr support is really convenient! I have not tested the mdim interface in sf/stars yet, but it's on my to-do list. Would you know about any good resources to draw some inspiration from?

**Reply by cboettig** [reply]
for zarr in R, https://r-spatial.org//r/2022/09/13/zarr.html is quite a good starting place I think.  The tough part for me is adapting to manually setting the `offset`, `step`, and `count` in `read_mdim` to read only the desired subset of data (in each dimension).  It's not so complicated I guess, but feels very low-level when I have become use to operations that are lazy by default.  e.g. instead of saying "I want data in this bounding box at this resolution" (a la gdalcubes), we manually say something like "I want the values between cell 59 and cell 157 in the x direction and y direction, for the last value of the "t" dimension and skip every other cell" -- and of course you have to provide (in order) this step, offset, and count for each of the dimensions of the object.  But I'm also a total novice here so perhaps there are easier things.  @mdsumner probably has some good suggestions.

**Reply by mdsumner** [reply]
tidync might work, or read_ncdf in stars with "#mode=zarr": https://gist.github.com/mdsumner/492d2a98bffc6de5974a96f50a0b75f2?permalink_comment_id=5071910#gistcomment-5071910

I started some work on mdim support in gdalraster but it will be a while before anything substantial from me there 🙏

**Reply by goergen95** [reply]
Thanks, @mdsumner! Will have a look! 

Looking at the r-spatial post, I'd guess its a question of 
lazyness and proper slicing methods, isn't it? 

Like, do you have more insights into why something like the
Python code from the post currently is not possible with stars?
Is it because of the underlying GDAL mdim raster model or something with
stars?

```python
import fsspec
import xarray
import zarr
m = fsspec.get_mapper("https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706")
ds = xarray.open_zarr(m)
ds.sel(time=slice("1948-01-01","1955-12-31")).to_zarr("./psl.zarr")
```

**Reply by goergen95** [reply]
I guess what I am asking is why is `stars` able to provide lazy evaluation for `read_stars()` but not for `read_mdim()`, and if we had this functionality, does it not come really close to what `xarray` provides (maybe short of a richer methods set and sophisticated chunking)?

What I managed to do so far is tricking `stars` into reading `zarr` lazily by reading one slice at a time. It feels terribly wrong to do so and the performance is really bad too (I suspect `stars` actually reads the complete array). But it comes close to how I would like to work with nd-arrays in R, that is with high-level methods for slicing/aggregation/apply on lazily-evaluated arrays. I currently lack the understanding of the low-level challenges to achieve something like this.

``` r
library(stars)
#> Loading required package: abind
#> Loading required package: sf
#> Linking to GEOS 3.12.1, GDAL 3.8.2, PROJ 9.3.1; sf_use_s2() is TRUE

src <- 'ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/"'
gdal_utils("info", src)
#> Driver: Zarr/Zarr
#> Files: none associated
#> Size is 512, 512
#> Subdatasets:
#>   SUBDATASET_1_NAME=ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/":/lat
#>   SUBDATASET_1_DESC=Array /lat
#>   SUBDATASET_2_NAME=ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/":/lon
#>   SUBDATASET_2_DESC=Array /lon
#>   SUBDATASET_3_NAME=ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/":/time
#>   SUBDATASET_3_DESC=Array /time
#>   SUBDATASET_4_NAME=ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/":/lat_bnds
#>   SUBDATASET_4_DESC=Array /lat_bnds
#>   SUBDATASET_5_NAME=ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/":/lon_bnds
#>   SUBDATASET_5_DESC=Array /lon_bnds
#>   SUBDATASET_6_NAME=ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/":/psl
#>   SUBDATASET_6_DESC=Array /psl
#>   SUBDATASET_7_NAME=ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706/":/time_bnds
#>   SUBDATASET_7_DESC=Array /time_bnds
#> Corner Coordinates:
#> Upper Left  (    0.0,    0.0)
#> Lower Left  (    0.0,  512.0)
#> Upper Right (  512.0,    0.0)
#> Lower Right (  512.0,  512.0)
#> Center      (  256.0,  256.0)
ts <- as.POSIXct(read_stars(paste0(src, ":/time"))$time[ ,1])
dsns <- sprintf(paste0(src, ":/psl:%s"), 1:20)
d <- read_stars(dsns, proxy = TRUE, along = "time")
d <- st_set_dimensions(d, "time", values = ts[1:20])
print(d)
#> stars_proxy object with 1 attribute in 20 file(s); showing the first 10 filenames
#> $attr
#>  [1] "[...]/psl:1"  "[...]/psl:2"  "[...]/psl:3"  "[...]/psl:4"  "[...]/psl:5" 
#>  [6] "[...]/psl:6"  "[...]/psl:7"  "[...]/psl:8"  "[...]/psl:9"  "[...]/psl:10"
#> 
#> dimension(s):
#>      from  to         offset   delta  refsys x/y
#> x       1 288         -0.625    1.25      NA [x]
#> y       1 192         -90.47  0.9424      NA [y]
#> time    1  20 1948-01-01 UTC 6 hours POSIXct

bb <- st_bbox(c(xmin = 10.125, ymin = 0.125, xmax = 70.125, ymax = 70.125))
d2 <- d[bb]
d2 <- aggregate(d2, by = "1 day", FUN =mean)
plot(d2, reset = FALSE)
```

![](https://i.imgur.com/7C1U7C1.png)

``` r
print(d2)
#> stars_proxy object with 1 attribute in 20 file(s); showing the first 10 filenames
#> $attr
#>  [1] "[...]/psl:1"  "[...]/psl:2"  "[...]/psl:3"  "[...]/psl:4"  "[...]/psl:5" 
#>  [6] "[...]/psl:6"  "[...]/psl:7"  "[...]/psl:8"  "[...]/psl:9"  "[...]/psl:10"
#> 
#> dimension(s):
#>      from  to         offset   delta  refsys x/y
#> x       9  57         -0.625    1.25      NA [x]
#> y      97 171         -90.47  0.9424      NA [y]
#> time    1  20 1948-01-01 UTC 6 hours POSIXct    
#> call_list:
#> [[1]]
#> aggregate(x = x, by = by, FUN = FUN)
#> attr(,".Environment")
#> <environment: 0x5ea22a1d4f00>
#> 
#> This object has pending lazy operations: dimensions as printed may not reflect this.
```

<sup>Created on 2024-06-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by cboettig** [reply]
yeah good question.  It's a bit out of my depth, but I think part of the challenge with doing lazy eval in this comes from lower in the stack than stars, specifically from the difference between the 'standard' gdal interface and the gdal mdim interface, and the underlying data structures they are built around.  gdal mdim models the an arbitrary n-dimensional cube used in HDF5 and then zarr (and xarray); i.e. it isn't explicitly spatial and doesn't privilege a notion of 2-D space in any particular way -- unlike the standard gdal interface (and hence the standard read_stars() interface that binds it), read_mdim() can't assume it has any of this information about these dimensions.  of course that could be provided by metadata, but zarr and hdf5 are agnostic about how such metadata is represented (i.e. there's no expectation that the data is even spatial in nature).  The discussion at https://discourse.pangeo.io/t/example-which-highlights-the-limitations-of-netcdf-style-coordinates-for-large-geospatial-rasters/4140 might be helpful.  

Of course this doesn't preclude the idea lazy operations, but it does mean that there's a completely different execution pipeline involved, and not one that stars provides us.  I think in stars::read_mdim we would do something like:


```r
library(stars)
dsn = 'ZARR:"/vsicurl/https://storage.googleapis.com/cmip6/CMIP6/HighResMIP/CMCC/CMCC-CM2-HR4/highresSST-present/r1i1p1f1/6hrPlev/psl/gn/v20170706"/:psl.zarr/'

# peek at the metadata without downloading:
ret = gdal_utils("mdiminfo", dsn, quiet = TRUE)
jsonlite::fromJSON(ret)$dimensions |> tibble::as_tibble()

# spatial subset with offset + count
d2 = read_mdim(dsn, offset = c(10,10,0), count = c(40, 40, 1)) 
st_crs(d2) = 'OGC:CRS84'
d2

# read every 4th pixel in x and y (downsample), and read only one value each year (every 6hr * 365days, note data ignores leap years):
d = read_mdim(dsn, step = c(4, 4, 365 * 4)) 
st_crs(d) = 'OGC:CRS84'
d
```

At this point we have a (in memory) `stars` object that we can then compute as usual with.  

Other than the call mdiminfo to peek at the metadata, this isn't lazy per se, but still embraces the ability to leverage the range request ability of not downloading or reading into RAM more data than we are interested in.

**Reply by cboettig** [reply]
on the down side, I don't see an obvious way to stream this other than manually, e.g. if you wanted the `d2` array but averaged over all time, without ever holding all those values in RAM at once, my understanding is that you would literally need a `for` loop over read_mdim().  @edzer might be able to correct me on that though.

**Reply by edzer** [reply]
Thanks for dragging me in this interesting issue!  @cboettig brought some of this up here: https://github.com/r-spatial/stars/issues/659 . Yes, the `read_mdim()` in stars is currently pretty minimal, reflecting what the underlying GDAL function can do: read an array, possibly starting from a certain offset, with a certain size, possibly skipping every n cells (is this called strided read?)  It would be nice if it read nothing, then only do everything you wish in a cleverly chunked way, like some of the methods for `stars_proxy` do. That problem is however pretty complicated, and will take some time to implement. Any help is welcome!

**Reply by cboettig** [reply]
Thanks @edzer!  Yes I think this mapme.biodiversity team has been doing some really cool stuff, and obviously leaning heavily on your work.

Maybe should move this over to the other thread but anyway:

Yeah, I suspected read_mdim was rather constrained by what the gdal mdim interface does and doesn't do.  As nice as the idea sounds in theory, it probably doesn't make sense to try and implement everything stars can do with the standard gdal interface over on the mdim side, given that even the OSGeo team hasn't attempted feature parity there.  

I do think @mdsumner 's instincts are right though in that we don't really have anything equivalent to xarray in R, i.e. a very intuitive, R-native or "tidy" implementation of (not-necessarily-spatial) n-dimensional array data.  Yeah, we have the various bindings to netcdf and hdf5 C libraries, and Mike's tidync might be closest, but nothing quite like xarray syntax like in the https://dynamical.org examples (or many others) up top.  On the flip side, I do think GDAL has optimized the hell out the virtual filesystem bindings, and I think at least some xarray pipelines pay a real overhead by trying to do file virtualization in fsspec etc (i.e. in python, rather than C).  Though I do kind of admire the way xarray syntax is agnostic (or nearly so) across some 5 or so different backend drivers, (including gdal if we go with rioxarray, though that mechanism seems remarkably un-loved by pythonistas).

**Reply by mdsumner** [reply]
VirtualiZarr (the natural successor of kerchunk) is the wildest thing I've found, and I think it has potential and I'm already using it to transcend from the old raster package stuff we built, it's essentially the same as VRT-for-xarray that I was long looking for. It uses references to store where array blobs live in file formats. I think it's a bit silly to use for tiff, but when you think like xarray it's natural to want to avoid GDAL and any actual formal library support. (Doesn't yet support HDF4 but HDF5, netcdf, tiff, and grib are). It obviously can't support some compression encodings without being a lot more complex, but it works great for those enormous global holdings of netcdf, hdf, grib. 👌

There's pizarrr for R, I don't know if there's a real C lib for Zarr but GDAL's inbuilt is independent.

GDAL multidim is a bit undercooked but it's really a matter of having more than one person actually contribute to it (community discourse could help here), and all of this is in that uncomfortable space where matlab-think meets geospatial. If xarray (and hence also Zarr) would support extent georeferencing and not just degenerate rectilinear coords it would be better. 

Python communities are pretty GDAL agnostic and that's a huge mistake. They only see it via rasterio which is a major problem. I guess if 'wheels' (binary package builds) were available for osgeo.gdal that would help, so that's something the R community could obviously help with. R now has good GDAL raster and vector API support in {gdalraster}. Get behind it!

**Reply by goergen95** [reply]
stars recently gained initial support for lazy-evaluation of [multidim data sets](https://github.com/r-spatial/stars/commit/ad4748fa0d13f31531e7ec9eb2899e0870748551), however, I tend to agree that GDAL's mdim-functionality currently feels undercooked and I am not sure how much this restricts what stars can/will be able to do.

I think I do see the path for a POC off of something like a virtual Zarr in R (though I am surely terribly underestimating the effort required to achieve this). I am unsure as to where is the best place to put something like this, too. Could stars (maybe eventually even GDAL) consume a virtual Zarr store? If so, we could draw on the methods already implemented there for actually getting things done with the underlying data. If not, one could build upon gdalraster, but substantial effort would be required to re-implement methods (such as subset, concatenation, aggregation, ...) on multidimensional, labelled arrays.

Naively sketching out the components I think would be required to get to a first POC:

- methods for ingesting multidim data (e.g. netcdf and other formats) into a virtual Zarr store:
  I imagine this to look something like a multidim [GTI](https://gdal.org/drivers/raster/gti.html) that references the chunks of multidim arrays.
  This potentially should be serialized to Parquet to allow for efficient queries when data actually needs to be read.

- call-stack mechanism:
  This mechanism should allow for the usual operations, similar to what stars already provides, e.g. cropping, sub-setting, aggregating, applying functions etc. This would also require to update dimensions/labels on-the-fly, like stars does, so you always know the shape of your arrays. I think this is actually the part that requires the most effort to get it right, but stars already has most of this, so it would make sense to build up on this.

- compute-stage:
  Data I/O via GDAL to read the required chunks as R arrays. Then apply the operations in the stack to those chunks. This could be parallelized via the future framework, allowing to scale from multi-core on a single machine to distributed compute nodes relatively cheap. I think this would also require an efficient mechanism to serialize (intermediate) results of chunks to disk.

**Reply by mdsumner** [reply]
I'm not across it enough where it does already work, but definitely interested to discuss the ideas. 

It's also not always as compelling as I've been led to believe, it's faster to open 15000 NetCDFs and pull out one pixel with terra+furrr than it is to open a kerchunk json version of the entire collection with dask (instant), sel() or isel() to that pixel and variable (also instant), and then call .values (takes ~5x longer than terra start to finish).  so not enough to say that dask isn't parallelizing, but I don't know how to isolate it to make a fair comparison yet.  I mostly want some real discussion and detailed examples, but switching languages still largely means not talking about the other alternatives in context.

GDAL is always lazy  - the only interface that's faithful is osgeo.gdal, and fwiw that is really the best way to learn the API, it's how most features are tested too. xarry already handles multiple engines that way, but it's not strong on leveraging GDAL either (rioxararay puts a particular lens on it, much like odc does too). but gdalraster is quickly becoming very rich in features, entirely suitable for extension and building tools with.

**Comment by karpfen** [comment]
I just stumbled upon https://overturemaps.org/. Right now, it looks a bit like a mirror for OSM data to me, but maybe this is something to keep an eye on, since it may evolve into a cool data source for non-remote sensing based indicators.

**Reply by goergen95** [reply]
Interesting! I assume most of the data form there is vector based, right? That would require the package to take next steps into handling vector data more efficiently, e.g. by going through [duckdb](https://github.com/cboettig/duckdbfs).

**Reply by cboettig** [reply]
That is quite cool!  Yup, looks like they are [using geoparquet](https://docs.overturemaps.org/getting-data/#geoparquet), and docs include [duckdb examples](https://docs.overturemaps.org/getting-data/duckdb/) as well as utilities for [pmtiles](https://docs.overturemaps.org/examples/overture-tiles/) for efficient visualization.

**Comment by goergen95** [comment]
[Paper](https://www.preprints.org/manuscript/202408.1146/v1) on the technical implementation of the DOPA processing back-end.

**Comment by goergen95** [comment]
Annual global total estimated wildfire carbon emissions from 2003 to 2023, Source: [CAMS](https://atmosphere.copernicus.eu/state-wildfires-2023-24-cams-data-supports-assessment?utm_source=socialmedia&utm_medium=li&utm_id=news-wildfire-cams-data-0824)

**Comment by goergen95** [comment]
R package [post](https://loreabad6.github.io/post/) to streamline vector datacubes for polygons that change over space and time.

**Reply by fBedecarrats** [reply]
Interesting!

**Comment by fBedecarrats** [comment]
3ie has launched a new [inventory of remote-sensing indicators](https://www.3ieimpact.org/resources/remote-sensing-inventory) for development, with [this blog](https://3ieimpact.org/blogs/explore-3ies-new-inventory-remote-sensing-indicators-leverage-satellite-data-development) providing some background. “If you are interested in using geospatial data, or are interpreting geospatial data, you can navigate here, to your sector of interest, and then be presented with a list of RS indicators and which social science variables they are used as a proxy for. You can find descriptions and pros and cons for each indicator, and then can explore the paper repository to find methodological and empirical papers related to that indicator” “The largest categories were agriculture (37 papers), firms (30), education (29), social protection (24), and macroeconomics (20).”.  He then provides one sentence summaries of the education papers.
[Source](https://blogs.worldbank.org/en/impactevaluations/weekly-links-april-4--indian-structural-transformation-from-wwii)

**Comment by fBedecarrats** [comment]
Interesting resource here: [rspatialdata is a collection of data sources and tutorials on visualising spatial data using R](https://rspatialdata.github.io/)

**Reply by zivankaraman** [reply]
@fBedecarrats, thank you for this interesting info.
